File: src/main/java/com/kingmang/lazurite/parser/AST/Statements/AssertStatement.java
package com.kingmang.lazurite.parser.AST.Statements;

import com.kingmang.lazurite.console.Console;
import com.kingmang.lazurite.exceptions.LzrException;
import com.kingmang.lazurite.parser.AST.Expressions.Expression;
import com.kingmang.lazurite.parser.AST.InterruptableNode;
import com.kingmang.lazurite.patterns.visitor.ResultVisitor;
import com.kingmang.lazurite.patterns.visitor.Visitor;
import lombok.AllArgsConstructor;

@AllArgsConstructor
public final class AssertStatement extends InterruptableNode implements Statement {

    public final Expression expression;

    @Override
    public void execute() {
        super.interruptionCheck();
        int expr = expression.eval().asInt();
        boolean booleanExpr;
        if(expr == 1)booleanExpr = true;
        else booleanExpr = false;

        if (!booleanExpr) {
            throw new LzrException("AssertionError", "Assertion failed: " + expression);
        }
    }

    @Override
    public void accept(Visitor visitor) {
        visitor.visit(this);
    }

    @Override
    public <R, T> R accept(ResultVisitor<R, T> visitor, T t) {
        //return null;
        return visitor.visit(this, t);
    }

    @Override
    public String toString() {
        return "assert " + expression;
    }
}

File: src/main/java/com/kingmang/lazurite/parser/ast/statements/AssertStatement.kt
package com.kingmang.lazurite.parser.ast.statements

import com.kingmang.lazurite.exceptions.LzrException
import com.kingmang.lazurite.parser.ast.InterruptableNode
import com.kingmang.lazurite.parser.ast.expressions.Expression
import com.kingmang.lazurite.patterns.visitor.ResultVisitor
import com.kingmang.lazurite.patterns.visitor.Visitor

data class AssertStatement(val expression: Expression) : InterruptableNode(), Statement {
    override fun execute() {
        super.interruptionCheck()
        if (this.expression.eval().asInt() != 1) {
            throw LzrException("AssertionError", "Assertion failed: ${this.expression}")
        }
    }

    override fun accept(visitor: Visitor) =
        visitor.visit(this)

    override fun <R, T> accept(visitor: ResultVisitor<R, T>, input: T): R? =
        visitor.visit(this, input)

    override fun toString(): String =
        "assert ${this.expression}"
}
